core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataMappingService.java

```diff
@@ -343,13 +338,16 @@ private ClusterState applyRequest(ClusterState currentState, PutMappingClusterSt
                 IndexMetaData.Builder indexMetaDataBuilder = IndexMetaData.builder(indexMetaData);
                 // Mapping updates on a single type may have side-effects on other types so we need to
                 // update mapping metadata on all types
-                for (DocumentMapper mapper : indexService.mapperService().docMappers(true)) {
+                for (DocumentMapper mapper : mapperService.docMappers(true)) {
                     indexMetaDataBuilder.putMapping(new MappingMetaData(mapper.mappingSource()));
                 }
                 builder.put(indexMetaDataBuilder);
             }
-
-            return ClusterState.builder(currentState).metaData(builder).build();
+            if (updated) {
+                return ClusterState.builder(currentState).metaData(builder).build();
+            } else {
+                return currentState;
+            }
```

Comment:
++ I think this can buy use quite a fair bit of processing savings... yet, I wonder if we can improve the CS builder to detect this automatically (for sure not here)?